library/core/src/ptr/non_null.rs

```diff
@@ -552,6 +554,10 @@ impl<T: ?Sized> NonNull<T> {
     #[must_use = "returns a new pointer rather than modifying its argument"]
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(count.checked_mul(core::mem::size_of::<T>()).is_some())] // Prevent offset overflow
+    #[requires(count * core::mem::size_of::<T>() <= isize::MAX as usize)] // SAFETY: count * size_of::<T>() does not overflow isize
+    //#[requires(ub_checks::can_write(self.as_ptr().offset(count as isize)))] // not working
```

Comment:
As mentioned in https://github.com/model-checking/verify-rust-std/discussions/99#discussioncomment-10871535, the solution is to combine this requires clause with the `count.checked_mul` one to make sure they're properly applied.

Reply:
resolved with `same_allocation` and checking wrapping add.